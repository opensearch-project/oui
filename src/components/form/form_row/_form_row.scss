/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/**
 * 1. Coerce inline form elements to behave as block-level elements.
 * 2. For inline forms, we need to add margin if the label doesn't exist.
 */
.ouiFormRow {
  display: flex; /* 1 */
  flex-direction: column; /* 1 */
  max-width: $ouiFormMaxWidth;

  + .ouiFormRow,
  + .ouiButton {
    margin-top: $ouiSize;
  }
}

.ouiFormRow--fullWidth {
  max-width: 100%;
}

.ouiFormRow--hasEmptyLabelSpace {
  margin-top: ($ouiFontSizeXS * $ouiLineHeight) + $ouiSizeXS; /* 2 */
  // the following ensure that contents that aren't inherently the same height
  // as inputs will align to the vertical center
  min-height: $ouiFormControlHeight;
  padding-bottom: 0;
  justify-content: center;
}

.ouiFormRow__labelWrapper {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  margin-bottom: $ouiSizeXS;
}

.ouiFormRow--horizontal {
  flex-direction: row;
  align-items: stretch;

  .ouiFormRow__label {
    @include ouiTextBreakWord;
    hyphens: auto;
    max-width: 100%; // Fixes IE
  }

  .ouiFormRow__labelWrapper {
    display: block;
    line-height: $ouiFormControlCompressedHeight - 1px; // The 1px less helps the alignment of the text baseline
    width: calc(33% - #{$ouiSizeS});
    margin-right: $ouiSizeS;
    margin-bottom: 0;
  }

  .ouiFormRow__fieldWrapper {
    width: 67%;
  }

  + .ouiFormRow--horizontal {
    margin-top: $ouiSizeS;
  }

  + .ouiFormRow--horizontal.ouiFormRow--hasSwitch {
    margin-top: $ouiSizeM; // More spacing since we reduced the height to match that of the switch
  }

  &.ouiFormRow--hasSwitch {
    .ouiFormRow__labelWrapper {
      line-height: $ouiSwitchHeight - 1px; // The 1px less helps the alignment of the text baseline
      width: auto;
      min-width: calc(33% - #{$ouiSizeS});
    }

    .ouiFormRow__fieldWrapper {
      width: auto;

      .ouiSwitch--compressed {
        margin-top: calc($ouiSizeXS / 2); // Better vertical alignment of a compressed switch to the horizontal label
      }
    }

    + .ouiFormRow--horizontal {
      margin-top: $ouiSizeM; // More spacing since we reduced the height to match that of the switch
    }
  }
}

.ouiFormRow__fieldWrapperDisplayOnly {
  min-height: $ouiFormControlHeight;
  display: flex;
  align-items: center;
}

.ouiFormRow--compressed {
  &.ouiFormRow--hasEmptyLabelSpace {
    min-height: $ouiFormControlCompressedHeight;
  }

  .ouiFormRow__fieldWrapperDisplayOnly {
    min-height: $ouiFormControlCompressedHeight;
  }
}
