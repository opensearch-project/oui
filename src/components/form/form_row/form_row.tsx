/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  cloneElement,
  Component,
  Children,
  HTMLAttributes,
  ReactElement,
  ReactNode,
} from 'react';
import classNames from 'classnames';
import { ExclusiveUnion, CommonProps, keysOf } from '../../common';

import { get } from '../../../services/objects';

import { OuiFormHelpText } from '../form_help_text';
import { OuiFormErrorText } from '../form_error_text';
import { OuiFormLabel } from '../form_label';

import { htmlIdGenerator } from '../../../services/accessibility';

const displayToClassNameMap = {
  row: null,
  rowCompressed: 'ouiFormRow--compressed',
  columnCompressed: 'ouiFormRow--compressed ouiFormRow--horizontal',
  center: null,
  centerCompressed: 'ouiFormRow--compressed',
  columnCompressedSwitch:
    'ouiFormRow--compressed ouiFormRow--horizontal ouiFormRow--hasSwitch',
};

export const DISPLAYS = keysOf(displayToClassNameMap);

export type OuiFormRowDisplayKeys = keyof typeof displayToClassNameMap;

interface OuiFormRowState {
  isFocused: boolean;
  id: string;
}

type OuiFormRowCommonProps = CommonProps & {
  /**
   * When `rowCompressed`, just tightens up the spacing;
   * Set to `columnCompressed` if compressed
   * and horizontal layout is needed.
   * Set to `center` or `centerCompressed` to align non-input
   * content better with inline rows.
   * Set to `columnCompressedSwitch` if the form control being passed
   * as the child is a switch.
   */
  display?: OuiFormRowDisplayKeys;
  hasEmptyLabelSpace?: boolean;
  fullWidth?: boolean;
  /**
   * IDs of additional elements that should be part of children's `aria-describedby`
   */
  describedByIds?: string[];
  /**
   * Escape hatch to not render duplicate labels if the child also renders a label
   */
  hasChildLabel?: boolean;
  /**
   * ReactElement to render as this component's content
   */
  children: ReactElement;
  label?: ReactNode;
  /**
   * Adds an extra node to the right of the form label without
   * being contained inside the form label. Good for things
   * like documentation links.
   */
  labelAppend?: any;
  id?: string;
  isInvalid?: boolean;
  error?: ReactNode | ReactNode[];
  /**
   *  Adds a single node/string or an array of nodes/strings below the input
   */
  helpText?: ReactNode | ReactNode[];
};

type LabelProps = {
  labelType?: 'label';
} & OuiFormRowCommonProps &
  HTMLAttributes<HTMLDivElement>;

type LegendProps = {
  /**
   * Defaults to rendering a `<label>` but if passed `'legend'` for labelType,
   * will render both a `<legend>` and the surrounding container as a `<fieldset>`
   */
  labelType?: 'legend';
} & OuiFormRowCommonProps &
  HTMLAttributes<HTMLFieldSetElement>;

export type OuiFormRowProps = ExclusiveUnion<LabelProps, LegendProps>;

export class OuiFormRow extends Component<OuiFormRowProps, OuiFormRowState> {
  static defaultProps = {
    display: 'row',
    hasEmptyLabelSpace: false,
    fullWidth: false,
    describedByIds: [],
    labelType: 'label',
    hasChildLabel: true,
  };

  state: OuiFormRowState = {
    isFocused: false,
    id: this.props.id || htmlIdGenerator()(),
  };

  onFocus = (...args: any[]) => {
    // Doing this to allow onFocus to be called correctly from the child input element as this component overrides it
    const onChildFocus = get(this.props, 'children.props.onFocus');
    if (onChildFocus) {
      onChildFocus(...args);
    }

    this.setState(({ isFocused }) => {
      if (!isFocused) {
        return {
          isFocused: true,
        };
      } else {
        return null;
      }
    });
  };

  onBlur = (...args: any[]) => {
    // Doing this to allow onBlur to be called correctly from the child input element as this component overrides it
    const onChildBlur = get(this.props, 'children.props.onBlur');
    if (onChildBlur) {
      onChildBlur(...args);
    }

    this.setState({
      isFocused: false,
    });
  };

  render() {
    const {
      children,
      helpText,
      isInvalid,
      error,
      label,
      labelType,
      labelAppend,
      hasEmptyLabelSpace,
      fullWidth,
      className,
      describedByIds,
      display,
      hasChildLabel,
      id: propsId,
      ...rest
    } = this.props;

    const { id } = this.state;

    const classes = classNames(
      'ouiFormRow',
      {
        'ouiFormRow--hasEmptyLabelSpace': hasEmptyLabelSpace,
        'ouiFormRow--fullWidth': fullWidth,
      },
      displayToClassNameMap[display!], // Safe use of ! as default prop is 'row'
      className
    );

    let optionalHelpTexts;

    if (helpText) {
      const helpTexts = Array.isArray(helpText) ? helpText : [helpText];
      optionalHelpTexts = helpTexts.map((helpText, i) => {
        const key = typeof helpText === 'string' ? helpText : i;
        return (
          <OuiFormHelpText
            key={key}
            id={`${id}-help-${i}`}
            className="ouiFormRow__text">
            {helpText}
          </OuiFormHelpText>
        );
      });
    }

    let optionalErrors;

    if (error && isInvalid) {
      const errorTexts = Array.isArray(error) ? error : [error];
      optionalErrors = errorTexts.map((error, i) => {
        const key = typeof error === 'string' ? error : i;
        return (
          <OuiFormErrorText
            key={key}
            id={`${id}-error-${i}`}
            className="ouiFormRow__text">
            {error}
          </OuiFormErrorText>
        );
      });
    }

    let optionalLabel;
    const isLegend = label && labelType === 'legend' ? true : false;

    if (label || labelAppend) {
      let labelProps = {};
      if (isLegend) {
        labelProps = {
          type: labelType,
        };
      } else {
        labelProps = {
          htmlFor: hasChildLabel ? id : undefined,
          isFocused: this.state.isFocused,
          type: labelType,
        };
      }
      optionalLabel = (
        <div className="ouiFormRow__labelWrapper">
          <OuiFormLabel
            className="ouiFormRow__label"
            isInvalid={isInvalid}
            aria-invalid={isInvalid}
            {...labelProps}>
            {label}
          </OuiFormLabel>
          {labelAppend && ' '}
          {labelAppend}
        </div>
      );
    }

    const optionalProps: React.AriaAttributes = {};
    /**
     * Safe use of ! as default prop is []
     */
    const describingIds = [...describedByIds!];

    if (optionalHelpTexts) {
      optionalHelpTexts.forEach((optionalHelpText) =>
        describingIds.push(optionalHelpText.props.id)
      );
    }

    if (optionalErrors) {
      optionalErrors.forEach((error) => describingIds.push(error.props.id));
    }

    if (describingIds.length > 0) {
      optionalProps['aria-describedby'] = describingIds.join(' ');
    }

    const field = cloneElement(Children.only(children), {
      id,
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      ...optionalProps,
    });

    const fieldWrapperClasses = classNames('ouiFormRow__fieldWrapper', {
      ouiFormRow__fieldWrapperDisplayOnly:
        /**
         * Safe use of ! as default prop is 'row'
         */
        display!.startsWith('center'),
    });

    const sharedProps = {
      className: classes,
      id: `${id}-row`,
    };

    const contents = (
      <React.Fragment>
        {optionalLabel}
        <div className={fieldWrapperClasses}>
          {field}
          {optionalErrors}
          {optionalHelpTexts}
        </div>
      </React.Fragment>
    );

    return labelType === 'legend' ? (
      <fieldset
        {...sharedProps}
        {...(rest as HTMLAttributes<HTMLFieldSetElement>)}>
        {contents}
      </fieldset>
    ) : (
      <div {...sharedProps} {...(rest as HTMLAttributes<HTMLDivElement>)}>
        {contents}
      </div>
    );
  }
}

// @internal
export type OuiCompressedFormRowProps = ExclusiveUnion<
  Omit<LabelProps, 'display'>,
  Omit<LegendProps, 'display'>
>;

// @internal
export class OuiCompressedFormRow extends OuiFormRow {
  static defaultProps = {
    ...OuiFormRow.defaultProps,
    display: 'rowCompressed',
  };
}
