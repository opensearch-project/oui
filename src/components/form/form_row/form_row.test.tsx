/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { requiredProps } from '../../../test';

import { OuiFormRow, DISPLAYS } from './form_row';

describe('OuiFormRow', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiFormRow {...requiredProps}>
        <input />
      </OuiFormRow>
    );

    expect(container).toMatchSnapshot();
  });

  test('ties together parts for accessibility', () => {
    const props = {
      label: 'Label',
      helpText: 'Help text',
      isInvalid: true,
      error: ['Error one', 'Error two'],
    };

    const { container } = render(
      <OuiFormRow {...requiredProps} {...props}>
        <input />
      </OuiFormRow>
    );

    const input = container.querySelector('input') as HTMLInputElement;
    const label = container.querySelector('label') as HTMLLabelElement;
    const helpText = container.querySelector('[id*="-help-"]') as HTMLElement;
    const errorTexts = container.querySelectorAll('[id*="-error-"]');

    // Input is labeled by the label.
    expect(input.id).toEqual('generated-id');
    expect(label.getAttribute('for')).toEqual('generated-id');

    // Input is described by help and error text.
    expect(helpText.id).toEqual('generated-id-help-0');
    expect(errorTexts[0].id).toEqual('generated-id-error-0');
    expect(errorTexts[1].id).toEqual('generated-id-error-1');
    expect(input.getAttribute('aria-describedby')).toEqual(
      'generated-id-help-0 generated-id-error-0 generated-id-error-1'
    );
  });

  describe('props', () => {
    test('label is rendered', () => {
      const { container } = render(
        <OuiFormRow label="label">
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('label append is rendered', () => {
      const { container } = render(
        <OuiFormRow label="label" labelAppend="append">
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('label renders as a legend and subsquently a fieldset wrapper', () => {
      const { container } = render(
        <OuiFormRow label="label" labelType="legend">
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('describedByIds is rendered', () => {
      const { container } = render(
        <OuiFormRow describedByIds={['generated-id-additional']}>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('id is rendered', () => {
      const { container } = render(
        <OuiFormRow id="id">
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('isInvalid is rendered', () => {
      const { container } = render(
        <OuiFormRow isInvalid label="label">
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('error as string is rendered', () => {
      const { container } = render(
        <OuiFormRow error="Error" isInvalid={true}>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('error as array is rendered', () => {
      const { container } = render(
        <OuiFormRow error={['Error', 'Error2']} isInvalid={true}>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('error is not rendered if isInvalid is false', () => {
      const { container } = render(
        <OuiFormRow error={['Error']} isInvalid={false}>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('helpText is rendered', () => {
      const { container } = render(
        <OuiFormRow helpText={<span>This is help text.</span>}>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('hasEmptyLabelSpace is rendered', () => {
      const { container } = render(
        <OuiFormRow hasEmptyLabelSpace>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    test('fullWidth is rendered', () => {
      const { container } = render(
        <OuiFormRow fullWidth>
          <input />
        </OuiFormRow>
      );

      expect(container).toMatchSnapshot();
    });

    describe('display type', () => {
      DISPLAYS.forEach((display) => {
        test(`${display} is rendered`, () => {
          const { container } = render(
            <OuiFormRow display={display}>
              <input />
            </OuiFormRow>
          );

          expect(container).toMatchSnapshot();
        });
      });
    });
  });

  describe('behavior', () => {
    describe('onFocus', () => {
      test('is called in child', () => {
        const focusMock = jest.fn();

        const { container } = render(
          <OuiFormRow label={<span>Label</span>}>
            <input onFocus={focusMock} />
          </OuiFormRow>
        );

        const input = container.querySelector('input') as HTMLInputElement;
        fireEvent.focus(input);

        expect(focusMock).toBeCalledTimes(1);

        // Ensure the focus event is properly fired on the parent
        // which will pass down to the OuiFormLabel
        expect(container).toMatchSnapshot();
      });

      test('works in parent even if not in child', () => {
        const { container } = render(
          <OuiFormRow label={<span>Label</span>}>
            <input />
          </OuiFormRow>
        );

        const input = container.querySelector('input') as HTMLInputElement;
        fireEvent.focus(input);

        // Ensure the focus event is properly fired on the parent
        // which will pass down to the OuiFormLabel
        expect(container).toMatchSnapshot();
      });
    });

    describe('onBlur', () => {
      test('is called in child', () => {
        const blurMock = jest.fn();

        const { container } = render(
          <OuiFormRow label={<span>Label</span>}>
            <input onBlur={blurMock} />
          </OuiFormRow>
        );

        const input = container.querySelector('input') as HTMLInputElement;
        fireEvent.blur(input);

        expect(blurMock).toBeCalledTimes(1);

        // Ensure the blur event is properly fired on the parent
        // which will pass down to the OuiFormLabel
        expect(container).toMatchSnapshot();
      });

      test('works in parent even if not in child', () => {
        const { container } = render(
          <OuiFormRow label={<span>Label</span>}>
            <input />
          </OuiFormRow>
        );

        const input = container.querySelector('input') as HTMLInputElement;
        fireEvent.blur(input);

        // Ensure the blur event is properly fired on the parent
        // which will pass down to the OuiFormLabel
        expect(container).toMatchSnapshot();
      });
    });
  });
});
