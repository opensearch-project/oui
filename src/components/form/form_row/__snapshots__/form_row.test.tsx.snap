// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiFormRow behavior onBlur is called in child 1`] = `
<OuiFormRow
  describedByIds={[]}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="ouiFormRow"
    id="generated-id-row"
  >
    <div
      className="ouiFormRow__labelWrapper"
    >
      <OuiFormLabel
        className="ouiFormRow__label"
        htmlFor="generated-id"
        isFocused={false}
        type="label"
      >
        <label
          className="ouiFormLabel ouiFormRow__label"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </OuiFormLabel>
    </div>
    <div
      className="ouiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</OuiFormRow>
`;

exports[`OuiFormRow behavior onBlur works in parent even if not in child 1`] = `
<OuiFormRow
  describedByIds={[]}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="ouiFormRow"
    id="generated-id-row"
  >
    <div
      className="ouiFormRow__labelWrapper"
    >
      <OuiFormLabel
        className="ouiFormRow__label"
        htmlFor="generated-id"
        isFocused={false}
        type="label"
      >
        <label
          className="ouiFormLabel ouiFormRow__label"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </OuiFormLabel>
    </div>
    <div
      className="ouiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</OuiFormRow>
`;

exports[`OuiFormRow behavior onFocus is called in child 1`] = `
<OuiFormRow
  describedByIds={[]}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="ouiFormRow"
    id="generated-id-row"
  >
    <div
      className="ouiFormRow__labelWrapper"
    >
      <OuiFormLabel
        className="ouiFormRow__label"
        htmlFor="generated-id"
        isFocused={true}
        type="label"
      >
        <label
          className="ouiFormLabel ouiFormRow__label ouiFormLabel-isFocused"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </OuiFormLabel>
    </div>
    <div
      className="ouiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</OuiFormRow>
`;

exports[`OuiFormRow behavior onFocus works in parent even if not in child 1`] = `
<OuiFormRow
  describedByIds={[]}
  display="row"
  fullWidth={false}
  hasChildLabel={true}
  hasEmptyLabelSpace={false}
  label={
    <span>
      Label
    </span>
  }
  labelType="label"
>
  <div
    className="ouiFormRow"
    id="generated-id-row"
  >
    <div
      className="ouiFormRow__labelWrapper"
    >
      <OuiFormLabel
        className="ouiFormRow__label"
        htmlFor="generated-id"
        isFocused={true}
        type="label"
      >
        <label
          className="ouiFormLabel ouiFormRow__label ouiFormLabel-isFocused"
          htmlFor="generated-id"
        >
          <span>
            Label
          </span>
        </label>
      </OuiFormLabel>
    </div>
    <div
      className="ouiFormRow__fieldWrapper"
    >
      <input
        id="generated-id"
        onBlur={[Function]}
        onFocus={[Function]}
      />
    </div>
  </div>
</OuiFormRow>
`;

exports[`OuiFormRow is rendered 1`] = `
<div
  aria-label="aria-label"
  class="ouiFormRow testClass1 testClass2"
  data-test-subj="test subject string"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props describedByIds is rendered 1`] = `
<div
  className="ouiFormRow"
  id="generated-id-row"
>
  <div
    className="ouiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-additional"
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`OuiFormRow props display type center is rendered 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper ouiFormRow__fieldWrapperDisplayOnly"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props display type centerCompressed is rendered 1`] = `
<div
  class="ouiFormRow ouiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper ouiFormRow__fieldWrapperDisplayOnly"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props display type columnCompressed is rendered 1`] = `
<div
  class="ouiFormRow ouiFormRow--compressed ouiFormRow--horizontal"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props display type columnCompressedSwitch is rendered 1`] = `
<div
  class="ouiFormRow ouiFormRow--compressed ouiFormRow--horizontal ouiFormRow--hasSwitch"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props display type row is rendered 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props display type rowCompressed is rendered 1`] = `
<div
  class="ouiFormRow ouiFormRow--compressed"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props error as array is rendered 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-error-0 generated-id-error-1"
      id="generated-id"
    />
    <div
      aria-live="polite"
      class="ouiFormErrorText ouiFormRow__text"
      id="generated-id-error-0"
    >
      Error
    </div>
    <div
      aria-live="polite"
      class="ouiFormErrorText ouiFormRow__text"
      id="generated-id-error-1"
    >
      Error2
    </div>
  </div>
</div>
`;

exports[`OuiFormRow props error as string is rendered 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-error-0"
      id="generated-id"
    />
    <div
      aria-live="polite"
      class="ouiFormErrorText ouiFormRow__text"
      id="generated-id-error-0"
    >
      Error
    </div>
  </div>
</div>
`;

exports[`OuiFormRow props error is not rendered if isInvalid is false 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props fullWidth is rendered 1`] = `
<div
  class="ouiFormRow ouiFormRow--fullWidth"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props hasEmptyLabelSpace is rendered 1`] = `
<div
  class="ouiFormRow ouiFormRow--hasEmptyLabelSpace"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props helpText is rendered 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      aria-describedby="generated-id-help-0"
      id="generated-id"
    />
    <div
      class="ouiFormHelpText ouiFormRow__text"
      id="generated-id-help-0"
    >
      <span>
        This is help text.
      </span>
    </div>
  </div>
</div>
`;

exports[`OuiFormRow props id is rendered 1`] = `
<div
  class="ouiFormRow"
  id="id-row"
>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props isInvalid is rendered 1`] = `
<div
  class="ouiFormRow"
  id="generated-id-row"
>
  <div
    class="ouiFormRow__labelWrapper"
  >
    <label
      aria-invalid="true"
      class="ouiFormLabel ouiFormRow__label ouiFormLabel-isInvalid"
      for="generated-id"
    >
      label
    </label>
  </div>
  <div
    class="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
    />
  </div>
</div>
`;

exports[`OuiFormRow props label append is rendered 1`] = `
<div
  className="ouiFormRow"
  id="generated-id-row"
>
  <div
    className="ouiFormRow__labelWrapper"
  >
    <OuiFormLabel
      className="ouiFormRow__label"
      htmlFor="generated-id"
      isFocused={false}
      type="label"
    >
      label
    </OuiFormLabel>
     
    append
  </div>
  <div
    className="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`OuiFormRow props label is rendered 1`] = `
<div
  className="ouiFormRow"
  id="generated-id-row"
>
  <div
    className="ouiFormRow__labelWrapper"
  >
    <OuiFormLabel
      className="ouiFormRow__label"
      htmlFor="generated-id"
      isFocused={false}
      type="label"
    >
      label
    </OuiFormLabel>
  </div>
  <div
    className="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</div>
`;

exports[`OuiFormRow props label renders as a legend and subsquently a fieldset wrapper 1`] = `
<fieldset
  className="ouiFormRow"
  id="generated-id-row"
>
  <div
    className="ouiFormRow__labelWrapper"
  >
    <OuiFormLabel
      className="ouiFormRow__label"
      type="legend"
    >
      label
    </OuiFormLabel>
  </div>
  <div
    className="ouiFormRow__fieldWrapper"
  >
    <input
      id="generated-id"
      onBlur={[Function]}
      onFocus={[Function]}
    />
  </div>
</fieldset>
`;
